File: app/src/main/java/com/huanchengfly/tieba/post/utils/StringUtil.java
package com.huanchengfly.tieba.post.utils;

import android.content.Context;
import android.graphics.drawable.Drawable;
import android.text.Spannable;
import android.text.SpannableString;
import android.text.TextPaint;
import android.text.TextUtils;
import android.widget.TextView;

import com.huanchengfly.tieba.post.R;
import com.huanchengfly.tieba.post.components.spans.EmotionSpanV2;

import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class StringUtil {
    public static SpannableString getEmotionContent(int emotion_map_type, final TextView tv, CharSequence source) {
        try {
            if (source == null) {
                return new SpannableString("");
            }
            SpannableString spannableString;
            if (source instanceof SpannableString) {
                spannableString = (SpannableString) source;
            } else {
                spannableString = new SpannableString(source);
            }
            String regexEmotion = EmotionUtil.getRegex(emotion_map_type);
            Pattern patternEmotion = Pattern.compile(regexEmotion);
            Matcher matcherEmotion = patternEmotion.matcher(spannableString);
            while (matcherEmotion.find()) {
                String key = matcherEmotion.group();
                int start = matcherEmotion.start();
                String group1 = matcherEmotion.group(1);
                Drawable emotionDrawable = EmotionManager.INSTANCE.getEmotionDrawable(EmotionManager.INSTANCE.getEmotionIdByName(group1));
                if (emotionDrawable != null) {
                    TextPaint paint = tv.getPaint();
                    int size = Math.round(-paint.ascent() + paint.descent());
                    EmotionSpanV2 span = new EmotionSpanV2(emotionDrawable, size);
                    spannableString.setSpan(span, start, start + key.length(), Spannable.SPAN_EXCLUSIVE_EXCLUSIVE);
                }
            }
            return spannableString;
        } catch (Exception e) {
            e.printStackTrace();
            SpannableString spannableString;
            if (source instanceof SpannableString) {
                spannableString = (SpannableString) source;
            } else {
                spannableString = new SpannableString(source);
            }
            return spannableString;
        }
    }

    public static String getUsernameString(Context context, String username, String nickname) {
        boolean showBoth = SharedPreferencesUtil.get(context, SharedPreferencesUtil.SP_SETTINGS).getBoolean("show_both_username_and_nickname", false);
        if (TextUtils.isEmpty(nickname)) {
            return TextUtils.isEmpty(username) ? "" : username;
        } else if (showBoth && !TextUtils.isEmpty(username) && !TextUtils.equals(username, nickname)) {
            return context.getString(R.string.username_both, nickname, username);
        }
        return nickname;
    }

    public static String getAvatarUrl(String portrait) {
        if (TextUtils.isEmpty(portrait)) {
            return "";
        }
        if (portrait.startsWith("http://tb.himg.baidu.com/sys/portrait/item/")) {
            return portrait;
        }
        return "http://tb.himg.baidu.com/sys/portrait/item/" + portrait;
    }
}

File: app/src/main/java/com/huanchengfly/tieba/post/utils/StringUtil.kt
package com.huanchengfly.tieba.post.utils

import android.content.Context
import android.text.Spannable
import android.text.SpannableString
import android.text.SpannableStringBuilder
import android.text.Spanned
import android.text.TextUtils
import android.text.style.ForegroundColorSpan
import android.widget.TextView
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.AnnotatedString
import androidx.compose.ui.text.SpanStyle
import androidx.compose.ui.text.buildAnnotatedString
import androidx.compose.ui.text.withStyle
import com.huanchengfly.tieba.post.R
import com.huanchengfly.tieba.post.components.spans.EmoticonSpanV2
import com.huanchengfly.tieba.post.ui.common.theme.utils.ThemeUtils
import com.huanchengfly.tieba.post.utils.EmoticonManager.getEmoticonDrawable
import com.huanchengfly.tieba.post.utils.EmoticonManager.getEmoticonIdByName
import java.util.regex.Pattern
import kotlin.math.roundToInt

object StringUtil {
    @JvmStatic
    fun getEmoticonContent(
        emoticon_map_type: Int,
        tv: TextView,
        source: CharSequence?
    ): SpannableString {
        return try {
            if (source == null) {
                return SpannableString("")
            }
            val spannableString: SpannableString = if (source is SpannableString) {
                source
            } else {
                SpannableString(source)
            }
            val regexEmoticon = EmoticonUtil.getRegex(emoticon_map_type)
            val patternEmoticon = Pattern.compile(regexEmoticon)
            val matcherEmoticon = patternEmoticon.matcher(spannableString)
            while (matcherEmoticon.find()) {
                val key = matcherEmoticon.group()
                val start = matcherEmoticon.start()
                val group1 = matcherEmoticon.group(1) ?: ""
                val emoticonDrawable = getEmoticonDrawable(tv.context, getEmoticonIdByName(group1))
                if (emoticonDrawable != null) {
                    val paint = tv.paint
                    val size = (-paint.ascent() + paint.descent()).roundToInt()
                    val span = EmoticonSpanV2(emoticonDrawable, size)
                    spannableString.setSpan(
                        span,
                        start,
                        start + key.length,
                        Spannable.SPAN_EXCLUSIVE_EXCLUSIVE
                    )
                }
            }
            spannableString
        } catch (e: Exception) {
            e.printStackTrace()
            val spannableString: SpannableString = if (source is SpannableString) {
                source
            } else {
                SpannableString(source)
            }
            spannableString
        }
    }

    @JvmStatic
    fun getUsernameString(context: Context, username: String, nickname: String?): CharSequence {
        val showBoth = context.appPreferences.showBothUsernameAndNickname
        if (TextUtils.isEmpty(nickname)) {
            return if (TextUtils.isEmpty(username)) "" else username
        } else if (showBoth && !TextUtils.isEmpty(username) && !TextUtils.equals(
                username,
                nickname
            )
        ) {
            val builder = SpannableStringBuilder(nickname)
            builder.append(
                "($username)",
                ForegroundColorSpan(ThemeUtils.getColorByAttr(context, R.attr.color_text_disabled)),
                Spanned.SPAN_EXCLUSIVE_EXCLUSIVE
            )
            return builder
        }
        return nickname ?: ""
    }

    fun getUsernameAnnotatedString(
        context: Context,
        username: String,
        nickname: String?,
        color: Color
    ): AnnotatedString {
        val showBoth = context.appPreferences.showBothUsernameAndNickname
        return buildAnnotatedString {
            if (showBoth && !nickname.isNullOrEmpty() && username != nickname) {
                append(nickname)
                withStyle(SpanStyle(color = color)) {
                    append("(${username})")
                }
            } else {
                append(nickname ?: username)
            }
        }
    }

    @JvmStatic
    fun getAvatarUrl(portrait: String?): String {
        if (portrait.isNullOrEmpty()) {
            return ""
        }
        return if (portrait.startsWith("http://") || portrait.startsWith("https://")) {
            portrait
        } else "http://tb.himg.baidu.com/sys/portrait/item/$portrait"
    }

    @JvmStatic
    fun getBigAvatarUrl(portrait: String?): String {
        if (portrait.isNullOrEmpty()) {
            return ""
        }
        return if (portrait.startsWith("http://") || portrait.startsWith("https://")) {
            portrait
        } else "http://tb.himg.baidu.com/sys/portraith/item/$portrait"
    }

    fun String.getShortNumString(): String {
        val long = toLongOrNull() ?: return ""
        return long.getShortNumString()
    }

    fun Int.getShortNumString(): String {
        return toLong().getShortNumString()
    }

    fun Long.getShortNumString(): String {
        val long = this
        return if (long > 9999) {
            val longW = long * 10 / 10000L / 10F
            if (longW > 999) {
                val longKW = longW.toLong() * 10 / 1000L / 10F
                "${longKW}KW"
            } else {
                "${longW}W"
            }
        } else {
            "$this"
        }
    }
}